application/src/main/scala/com/azavea/franklin/api/services/TileService.scala

```diff
@@ -0,0 +1,86 @@
+package com.azavea.franklin.api.services
+
+import cats.data.Validated.{Invalid, Valid}
+import cats.data._
+import cats.effect._
+import cats.implicits._
+import com.azavea.franklin.api.endpoints._
+import com.azavea.franklin.database.StacItemDao
+import com.azavea.franklin.error.{NotFound => NF}
+import com.azavea.franklin.tile._
+import doobie._
+import doobie.implicits._
+import geotrellis.raster._
+import geotrellis.server.LayerTms
+import geotrellis.server._
+import org.http4s.HttpRoutes
+import org.http4s.dsl.Http4sDsl
+import sttp.tapir.server.http4s._
+
+class TileService[F[_]: Sync: LiftIO](enableTiles: Boolean, xa: Transactor[F])(
+    implicit cs: ContextShift[F],
+    csIO: ContextShift[IO]
+) extends Http4sDsl[F] {
+
+  import CogAssetNodeImplicits._
+
+  val tileEndpoints = new TileEndpoints(enableTiles)
+
+  def getTile(tileRequest: TileRequest): F[Either[NF, Array[Byte]]] = {
+    val assetKey     = tileRequest.asset
+    val collectionId = tileRequest.collection
+    val itemId       = tileRequest.item
+    val (z, x, y)    = tileRequest.zxy
+
+    val tileEither = for {
+      item <- EitherT.fromOptionF(
```

Comment:
`EitherT` worked out _really_ well here I think - being able to bail out at each part and be able to describe what went wrong looks nice.